Strato 2004 Add contract versioning

luketollefson opened the pull request:
https://blockapps.atlassian.net/browse/STRATO-2004

https://jenkins.blockapps.net/job/STRATO_test/4499/

**Note: this feature is only for SolidVM. EVM is not supported, nor do we have plans to support it.**

Add contract versioning based off the user's organization. For example, if a user was part of `bayer`, then any contract `Hello` they post will internally be `bayer:Hello`. This lets users of different organizations post contracts of the same name. So another user part of `blockapps` could post another contract named `Hello`, but internally the data will be saved to a table named `blockapps:Hello`. So then `bayer:Hello` and `blockapps:Hello` and coexist even if they are identical.

If the user is not part of any organization, then the contracts posted will not be part of a particular organization, e.g. contract `Hello` will be represented internally by a table named `Hello` (not `Hello`, etc.).

Additionally, this also supports contract versioning based off an application. So if the contract `Factory` spawns a new contract `Contract`, it would show up as `Factory:Contract`. If you were part of `blockapps`, it would be in the table `blockapps:Factory:Contract`.

To test out this feature, checkout "Try this feature out" at https://blockapps.atlassian.net/browse/STRATO-2004.

Commit:
Add contract versioning based on tx.organization

Commit:
Merge branch 'develop' into STRATO-2004-Add-contract-versioning

Commit:
Merge branch 'develop' into STRATO-2004-Add-contract-versioning

Commit:
Fix failing tests

Commit:
Use Address instead of Account to lookup X509Certs

Commit:
Add support for contract versioning based off application

Commit:
Fix failing tests

Commit:
Change indexing of certificates by Address instead of Account

dpmesko committed:
support in old bloc and new bloc for uploading contracts with codePtrs

dpmesko committed:
support in old bloc and new bloc for uploading contracts with codePtrs

dpmesko committed:
blockapps-rest allows codePtr in contract creation

dpmesko committed:
blockapps-rest allows codePtr in contract creation

dpmesko committed:
make codePtr optional

dpmesko committed:
make codePtr optional

dpmesko committed:
export/import CodePtr definition

dpmesko committed:
export/import CodePtr definition

dpmesko committed:
Merge branch 'STRATO-2008_codePtrContractCreations' into STRATO-2004-Add-contract-versioning

Commit:
Fix app versioning

Commit:
Add app versioning to events

Commit:
Change deliminator from / to :

Commit:
Pass tests with : version deliminator

Commit:
ACTUALLY have app versioning work!

Whew that was close! D:

dpmesko committed:
Merge branch 'STRATO-2004-Add-contract-versioning' of https://github.com/blockapps/strato-platform into STRATO-2004-Add-contract-versioning

Commit:
Attempt to fix issues in Jenkins test

Commit:
Merge branch 'develop' into STRATO-2004-Add-contract-versioning

dpmesko committed:
Merge branch 'STRATO-2004-Add-contract-versioning' of https://github.com/blockapps/strato-platform into STRATO-2004-Add-contract-versioning

Commit:
Merge branch 'STRATO-2004-Add-contract-versioning' into STRATO-2010-Add-certificates-to-LevelDB-once-block-is-built

Commit:
Add a cache for X509 certs

X509 certs will first go into a block cache if the TX is successful,
then only if the block is successful and is added will X509 certs be
placed in LevelDB.

dpmesko reviewed (changes requested):
I requested some changes around the `resolveCodePtr` bits, but otherwise it looks good to me!

Commit:
Add more unit tests for X509 versioning

dpmesko committed:
[WIP] expansion logic for history tables too, reworked globals

dpmesko committed:
[WIP] expansion logic for history tables too, reworked globals

dpmesko committed:
added event table expansion logic, reworked event emissions in SolidVM to get field name information for insertions

dpmesko committed:
added event table expansion logic, reworked event emissions in SolidVM to get field name information for insertions

dpmesko committed:
Merge branch 'STRATO-2017_expandCirrusTables' of https://github.com/blockapps/strato-platform into STRATO-2017_expandCirrusTables

dpmesko committed:
remove codePtr stuff I merged in

dpmesko committed:
Revert "support in old bloc and new bloc for uploading contracts with codePtrs"

This reverts commit f2117d220777797a61cf1682c4230c05986e70d8.

dpmesko committed:
Revert "make codePtr optional"

This reverts commit c2ef1c5fabdb763eea47f45d852e0c3413643246.

dpmesko committed:
Revert "export/import CodePtr definition"

This reverts commit 51f4702100a462245efe1d5786b0dae415a2b04e.

dpmesko committed:
Revert "blockapps-rest allows codePtr in contract creation"

This reverts commit 0ba4f3cf107e897be6bd75270abae92152ae04b3.

dpmesko committed:
fixed slipstream tests with new Event def

Commit:
Fix issue with cert cache

There was an issue where if certs were if in a block with two TXs
[TX1, TX2], where TX2 would register a cert and TX1 would "use" that
cert we would get a stateroot mismatch. This commit fixes that.

dustinnorwood committed:
Merge branch 'STRATO-2004-Add-contract-versioning' into STRATO-2010-Add-certificates-to-LevelDB-once-block-is-built

dustinnorwood committed:
Merge pull request #1209 from blockapps/STRATO-2010-Add-certificates-to-LevelDB-once-block-is-built

Strato 2010 Add Certificates to LevelDB Once Block is Built

dustinnorwood committed:
Merge branch 'develop' into STRATO-2004-Add-contract-versioning

dustinnorwood committed:
Remove redundant import of Control.DeepSeq from BlockApps.X509.Certificate

dustinnorwood reviewed (changes requested):
Great work - I'm requesting changes due to the spots where there are missing lookups from the in-memory certDB, which I believe are bugs. It would be nice to do the cleanup before merging, but I will approve the PR without it if the other bugs are fixed and the cleanup is holding up other work needed for the release.

dpmesko committed:
removed redefinition of CodePtr since it is replaced by TableName

dpmesko committed:
put the : even if org/app param is missing, to prevent table aliases; remove extra log statement

Commit:
Code cleanup

dpmesko committed:
if no org is present, the tablename is not parametrized

dpmesko committed:
everywhere we generate table names (even for index_ or _pkey), we use tableNameToText so that we can make changes in one place

dpmesko committed:
merge with master 2004, add pairs event Emission logic

dpmesko committed:
Merge pull request #1211 from blockapps/STRATO-2017_expandCirrusTables

[STRATO-2017] expand Cirrus tables

dpmesko committed:
non-escaped quotes in contract table to keep backwards compatibility
